### PR TITLE
refactor: deprecatedWarning

### PR DESCRIPTION
**Pull request by kiner-tang**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [X] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |   -  |
| 🇨🇳 Chinese |   -   |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at b6b1f90</samp>

This pull request refactors the warning logic for deprecated props in various components. It adds a new `deprecatedWarning` function in `components/_util/warning.ts` that generates consistent and clear warning messages. It replaces the `warning` function with the `deprecatedWarning` function in the components that have deprecated props, such as `Alert`, `Anchor`, `AutoComplete`, `BackTop`, `Breadcrumb`, `Calendar`, `Cascader`, `CollapsePanel`, `ConfigProvider`, `DatePicker`, `Drawer`, and `Dropdown`. It also updates the test cases in `components/auto-complete/__tests__/index.test.tsx`, `components/back-top/__tests__/index.test.tsx`, `components/drawer/__tests__/DrawerEvent.test.tsx`, and `components/dropdown/__tests__/index.test.tsx` to match the new warning message format.

### <samp>🤖 Generated by Copilot at b6b1f90</samp>

*  Add a new type and function to handle deprecated props in `warning.ts` ([link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-75853a37744088ef730adfb0906473685c80080827a097c12079581f39203ffbR60-R75))
* Replace `warning` function calls with `deprecatedWarning` function calls in various components, passing the component name, condition, old prop, and new prop as arguments ([link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-18c21d0f3bf7f7e9f29dd782a9b585a896225e10fcc6d470f4de65e822de794cL123-R126), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-353746edad3f07acb175d7e32c64972b87359b5a0dc4e14e4d3f1133b61e48a8L137-R138), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-c55203058d3a4064d8f52a5b4e7d7770153d215966d276d6814011d99268f76dL115-R117), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-c55203058d3a4064d8f52a5b4e7d7770153d215966d276d6814011d99268f76dL130-R126), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-22587f2ed95b1c480472e426d68bdb6cc32d38b3134e9aa5a32101b973312618L53-R55), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-f2426e3ced2642a4e369433c0c19b385fca436f57ad4620b4bd2bfcd7ab9254dL103-R113), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-a6c84736ab41b111c92cd33d3ae19e5b651015b5d09876c4295859d2abc323d0L45-R47), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-841e97dfb23daba72a34ee15810c8b36da6464eba16e120739ec6c9636f94fb2L128-R133), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-4737f6ef564486c449ec019f2826176d8b8d186088e5e912ad06ad1919b2d804L198-R200), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-7b26d89bc5d6c4b0c3933b263e17bb939b25f88b0e51e76014a3a29ae22ca223L28-R30), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-77b3e18cb24de05844f60221d8618bf867dd6d7186f601b9abb28a10db361617L14-R19), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-fba17c237c150271f91ee2ca848654484ecbbbcc71b010a8947620e945e10e34L81-R83), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-678a0178404ab41221775c01d203e79ccb84496e2f0b7cf4ef38222d7a7b562bL118-R119), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-10ffbf23ed7075c38649dfd0776998a514a56e42e09a91306f89b25e67e07d30L97-R98), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-978b29ecd5429963a073f5adfa7ded79d88d23850853c9fb6d453a82047365beL115-R123), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-978b29ecd5429963a073f5adfa7ded79d88d23850853c9fb6d453a82047365beL174-R166), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-9eb09be52eccb23eca0f8a281decb1e31a260e19b5176de99aef059c5abfb3c6L53-R55))
* Add `deprecatedWarning` import to the files that use the new function ([link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-18c21d0f3bf7f7e9f29dd782a9b585a896225e10fcc6d470f4de65e822de794cL13-R13), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-353746edad3f07acb175d7e32c64972b87359b5a0dc4e14e4d3f1133b61e48a8L8-R8), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-c55203058d3a4064d8f52a5b4e7d7770153d215966d276d6814011d99268f76dL10-R10), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-22587f2ed95b1c480472e426d68bdb6cc32d38b3134e9aa5a32101b973312618L11-R11), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-f2426e3ced2642a4e369433c0c19b385fca436f57ad4620b4bd2bfcd7ab9254dL8-R8), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-a6c84736ab41b111c92cd33d3ae19e5b651015b5d09876c4295859d2abc323d0L3-R3), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-4737f6ef564486c449ec019f2826176d8b8d186088e5e912ad06ad1919b2d804L23-R23), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-7b26d89bc5d6c4b0c3933b263e17bb939b25f88b0e51e76014a3a29ae22ca223L5-R5), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-678a0178404ab41221775c01d203e79ccb84496e2f0b7cf4ef38222d7a7b562bL14-R14), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-10ffbf23ed7075c38649dfd0776998a514a56e42e09a91306f89b25e67e07d30L9-R9), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-978b29ecd5429963a073f5adfa7ded79d88d23850853c9fb6d453a82047365beL14-R14))
* Replace `devUseWarning` import with `deprecatedWarning` import in the files that no longer use the former function ([link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-841e97dfb23daba72a34ee15810c8b36da6464eba16e120739ec6c9636f94fb2L13-R13), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-77b3e18cb24de05844f60221d8618bf867dd6d7186f601b9abb28a10db361617L3-R3), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-fba17c237c150271f91ee2ca848654484ecbbbcc71b010a8947620e945e10e34L13-R13), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-9eb09be52eccb23eca0f8a281decb1e31a260e19b5176de99aef059c5abfb3c6L5-R5))
* Create a reusable `deprecatedWarningFn` variable for the `Drawer` and `Dropdown` components, assigning the `deprecatedWarning` function call with the component name as 'Drawer' or 'Dropdown' ([link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-10ffbf23ed7075c38649dfd0776998a514a56e42e09a91306f89b25e67e07d30R92), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-978b29ecd5429963a073f5adfa7ded79d88d23850853c9fb6d453a82047365beR151))
* Remove commas after the words 'deprecated' in the expected warning messages in the test files to match the updated format ([link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-280a35bf9505afb3855621cda2722a2aae90b25ec454cd1a17ed3963a6714c19L113-R113), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-ca7e114756c8e1ba158e4d400f1d9b5dd5447095bb6274392c113e9658af831eL50-R50), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-6a54c10f458774e64ba74e7b984acec2119f6ca3c31893ec5cc3ed5d5ceb53c0L138-R141), [link](https://github.com/ant-design/ant-design/pull/44791/files?diff=unified&w=0#diff-a5e933ff15a7392fe4b00b0eb3b62fc823173b3ec9c87dda164f0e03488cf738L234-R237))

### CODE REVIEW

File: components/_util/warning.ts [hunk 1]
```diff
@@ -57,4 +57,20 @@ export const devUseWarning: () => TypeWarning =
       }
     : () => noop;
 
+export type DeprecatedWarning = (
+  component: string,
+) => (valid: boolean, oldProp: string, newProp: string) => void;
+
+export const deprecatedWarning: DeprecatedWarning = (component: string) => {
```
Comment (zombieJ) [inline components/_util/warning.ts:64]: 这个名字不太对，`deprecatedWarning` 看不出它是个 hook

File: components/_util/warning.ts [hunk 2]
```diff
@@ -57,4 +57,20 @@ export const devUseWarning: () => TypeWarning =
...
+  const warningFn = devUseWarning();
+  return (valid: boolean, oldProp: string, newProp: string) => {
+    warningFn(
+      valid,
+      component,
```
Comment (zombieJ) [inline components/_util/warning.ts:69]: `componentName` 看起来可以直接挪到 `devUseWarning` 上去

File: components/_util/warning.ts (hunk 1 above)
Comment (kiner-tang) [reply components/_util/warning.ts:64]: deprecatedWarning 本来就不是 hook 呀，只是一个工具方法

**Comment by afc163** [comment]
为啥变大了

**Comment by kiner-tang** [comment]
> 为啥变大了

增加了 deprecated 的，调用的时候方便一点

File: components/_util/warning.ts [hunk 3]
```diff
@@ -42,19 +46,34 @@ export const WarningContext = React.createContext<WarningContextProps>({});
...
 
-        const typeWarning: TypeWarning = (valid, component, type, message) => {
+        const typeWarning: TypeWarning = (valid, type, message) => {
           if (deprecated !== false || type !== 'deprecated') {
             warning(valid, component, message);
           }
         };
 
-        return typeWarning;
+        return {
```
Comment (zombieJ) [inline components/_util/warning.ts:60]: 返回的是一个 function，然后可以 `warn.deprecated`。而不是返回两个 function 的 obj

**Review by zombieJ — APPROVED** [review]

**Pull request merged**